libnetwork/etchosts/hosts.go

```diff
@@ -0,0 +1,202 @@
+		return fmt.Errorf("failed to create new hosts file: %w", err)
+	}
+	return nil
+}
+
+// Add the given entries to the hosts file, entries are only added if they are
+// not already present.
+// Add is not atomic because it will keep the current file inode. This is
+// required to keep bind mounts for containers working.
+func Add(file string, entries HostEntries) error {
+	if err := new(file, false, nil, entries, "", file); err != nil {
+		return fmt.Errorf("failed to add entry to hosts file: %w", err)
+	}
+	return nil
+}
+
+// new see comment on New()
+// addLocalhost is a extra bool to add localhost (if not already present)
+func new(baseFile string, addLocalhost bool, extraHosts []string, containerIPs HostEntries, hostContainersInternalIP, targetFile string) error {
+	entries, err := parseExtraHosts(extraHosts)
+	if err != nil {
+		return err
+	}
+	entries2, err := parseHostsFile(baseFile)
+	if err != nil {
+		return err
+	}
+
+	if err := writeHostFile(targetFile, append(entries, entries2...), containerIPs, hostContainersInternalIP, addLocalhost); err != nil {
+		return err
+	}
+	return nil
+}
+
+// parseExtraHosts converts a slice of "name:ip" string to entries.
+// Because podman and buildah both store the extra hosts in this format
+// we convert it here instead of having to this on the caller side.
+func parseExtraHosts(extraHosts []string) (HostEntries, error) {
+	entries := make(HostEntries, 0, len(extraHosts))
+	for _, entry := range extraHosts {
+		values := strings.SplitN(entry, ":", 2)
+		if len(values) != 2 {
+			return nil, fmt.Errorf("unable to parse host entry %q: incorrect format", entry)
+		}
+		if values[0] == "" {
+			return nil, fmt.Errorf("hostname in host entry %q is empty", entry)
+		}
+		if values[1] == "" {
+			return nil, fmt.Errorf("IP address in host entry %q is empty", entry)
+		}
+		e := HostEntry{IP: values[1], Names: []string{values[0]}}
+		entries = append(entries, e)
+	}
+	return entries, nil
+}
+
+// parseHostsFile parses a given host file and return s all entries in it.
+// Note that this will remove all comments and spaces.
+func parseHostsFile(file string) (HostEntries, error) {
+	// empty file is valid, in this case we skip adding entries from the file
+	if file == "" {
+		return nil, nil
+	}
+
+	f, err := os.Open(file)
+	if err != nil {
+		// do not error when the default hosts file does not exists
+		// https://github.com/containers/podman/issues/12667
+		if errors.Is(err, os.ErrNotExist) && file == DefaultHostsFile {
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer f.Close()
+
+	entries := HostEntries{}
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		// split of the comments
+		line := strings.SplitN(scanner.Text(), "#", 2)[0]
+		line = strings.TrimSpace(line)
+		if line == "" {
+			continue
+		}
+		fields := strings.Fields(line)
+		// if we only have a ip without names we skip it
```

Comment:
Since `strings.Fields` eats all the extra spaces around the fields, `strings.Trim` is redundant here.